### DISCUSSION TITLE
easy way to specify different paths for uprobe / #define ?

### DISCUSSION THREAD

**Discussion by laf0rge** [bpftrace/bpftrace#2218] [General]
I am frequently facing the problem that I have a bpftrace program which I would like to run on different systems, but "of course" the path name to the binary program or the shared library in question are different, requiring the manual adjustment of every single uprobe from something like uprobe:/usr/bin/foo to uprobe:/usr/local/bin/foo

I thought the obvious thing is to use #define for this, but somehow it doesn't work like this:

<pre>
#define OSMO_MSC "/usr/bin/osmo-msc"

uprobe:OSMO_MSC:my_symbol_name {
}
</pre>

as I'm getting
<pre>
./subsequent-poll.bt:52-56: ERROR: uprobe target file 'OSMO_MSC' does not exist or is not executable
</pre>

How can I pass the fully-qualified path for uprobe applications or libraries from either a variable/preprocessor definitiion or from an argument passed to bpftrace, or an environment variable ?

**Comment by viktormalik** [comment]
Hi, I have recommendations here:
- For libraries, you can use the library name, e.g. `uprobe:libc:malloc { ...}` and bpftrace will automatically determine the path
- For binaries, it's not that easy. `#define` directives cannot be used in probe names (only the program gets preprocessed), but you can use positional parameters instead:
```
bpftrace -e 'uprobe:$1:my_symbol_name { ... }' /usr/bin/osmo-msc
```
